File: src/main/java/com/sea/dsl/explain/DramaObject.java
package com.sea.dsl.explain;

public class DramaObject {
    private int index;
    private Context context;

    public DramaObject(Context context) {
        this.context = context;
    }

    public Context getContext() {
        return context;
    }

    public void setIndex(int index) {
        this.index = index;
    }

    public int getIndex() {
        return index;
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof DramaObject && this.index == ((DramaObject) obj).index;
    }

    @Override
    public int hashCode() {
        return index;
    }

    @Override
    public String toString() {
        return String.format("%s[%d]", this.getClass().getName(), index);
    }
}

File: src/main/scala/com/sea/dsl/explain/DramaObject.scala
package com.sea.dsl.explain

import org.apache.spark.sql.Column

class DramaObject(val column: Column) {


  def plus(any: Any): DramaObject = {
    apply(any, DramaObject.ops("+"))
  }

  def gt(any: Any): DramaObject = {
    apply(any, DramaObject.ops(">"))
  }

  def eq(any: Any): DramaObject = {
    apply(any, DramaObject.ops("==="))
  }

  def apply(any: Any, f: (Column, Any) => Column) = {
    any match {
      case dramaObject: DramaObject => DramaObject(f(this.column, dramaObject.column))
      case _ => DramaObject(f(this.column, any))
    }
  }

  override def toString: String ={
    column.toString()
  }
}

object DramaObject {
  def apply(column: Column): DramaObject = new DramaObject(column)

  def ops: Map[String, (Column, Any) => Column] = Map(
    "+" -> ((a, b) => a + b),
    "===" -> ((a, b) => a === b),
    ">" -> ((a, b) => a > b)
  )
}